File: examples/python/09-testing/main.py
# Testing in Python
# =====================
# unittest, pytest, doctest

import unittest
import pytest


# ── Basic unittest test ───────────────────────────

class TestBasic(unittest.TestCase):

    def test_addition(self):
        self.assertEqual(5 + 3, 8)

    def test_string_operations(self):
        s = "hello"
        self.assertEqual(len(s), 5)
        self.assertEqual(s.upper(), "HELLO")

    def test_list_operations(self):
        nums = [1, 2, 3]
        self.assertEqual(sum(nums), 6)
        self.assertEqual(len(nums), 3)


# ── Exception testing ──────────────────────────────

class TestExceptions(unittest.TestCase):

    def test_division_by_zero(self):
        with self.assertRaises(ZeroDivisionError):
            _ = 10 / 0

    def test_key_error(self):
        d = {"a": 1}
        with self.assertRaises(KeyError):
            _ = d["b"]


# ── Fixtures ──────────────────────────────────────

class TestFixtures(unittest.TestCase):

    def setUp(self):
        self.data = [1, 2, 3, 4, 5]

    def tearDown(self):
        self.data = []

    def test_with_fixture(self):
        self.assertEqual(sum(self.data), 15)


# ── Parameterized test (pytest) ──────────────────────

@pytest.mark.parametrize("input,expected", [
    (1, 1),
    (2, 4),
    (3, 9),
    (-5, 25),
])
def test_squares(input, expected):
    assert input * input == expected


# ── Pytest fixtures ─────────────────────────────────

@pytest.fixture
def sample_data():
    return {"name": "Alice", "age": 30}


def test_with_fixture(sample_data):
    assert sample_data["name"] == "Alice"
    assert sample_data["age"] == 30


# ── Mocking example ─────────────────────────────────

from unittest.mock import Mock, patch


class TestMocking(unittest.TestCase):

    def test_mock_function(self):
        mock_func = Mock(return_value=42)
        result = mock_func()
        self.assertEqual(result, 42)
        mock_func.assert_called_once()

    @patch("builtins.print")
    def test_mock_print(self, mock_print):
        print("test message")
        mock_print.assert_called_with("test message")


# ── Doctest ───────────────────────────────────────

def multiply(a, b):
    """
    Multiply two numbers.

    >>> multiply(2, 3)
    6
    >>> multiply(-2, 3)
    -6
    >>> multiply(0, 5)
    0
    """
    return a * b


# ── Summary test ─────────────────────────────────────

class TestSummary(unittest.TestCase):

    def test_python_testing_features(self):
        self.assertTrue(True)  # Always passes
        self.assertFalse(False)


# ── Running tests ───────────────────────────────────
# Run with:
# python -m unittest
# pytest
# pytest --cov=mymodule --cov-report=html
# python -m doctest main.py -v

File: examples/java/09-testing/main.java
// Testing in Java
// ===============
// JUnit, assertions, mocks

import org.junit.jupiter.api.*;
import static org.junit.jupiter.api.Assertions.*;
import org.junit.jupiter.params.*;
import org.junit.jupiter.params.provider.*;

class Main {
    // ── Basic assertion ─────────────────────────────

    @Test
    void testAddition() {
        int result = 5 + 3;
        assertEquals(8, result, "5 + 3 should be 8");
    }

    @Test
    void testStringEquals() {
        String actual = "hello";
        assertEquals("hello", actual);
    }

    // ── Exception testing ───────────────────────────

    @Test
    void testDivisionByZero() {
        ArithmeticException ex = assertThrows(
            ArithmeticException.class,
            () -> { int x = 10 / 0; }
        );
        assertEquals("/ by zero", ex.getMessage());
    }

    // ── Parameterized test ───────────────────────────

    @ParameterizedTest
    @CsvSource({"1,1", "2,4", "3,9", "5,25"})
    void testSquares(int input, int expected) {
        assertEquals(expected, input * input);
    }

    // ── Test lifecycle ──────────────────────────────

    private int counter;

    @BeforeEach
    void setup() {
        counter = 0;
        System.out.println("Before each test");
    }

    @Test
    void testCounter() {
        counter++;
        assertEquals(1, counter);
    }

    @AfterEach
    void teardown() {
        counter = 0;
        System.out.println("After each test");
    }

    // ── Assertions overview ───────────────────────────

    @Test
    void testAssertionTypes() {
        // Equality
        assertEquals("text", "text");
        assertNotEquals("text", "other");

        // Null
        assertNull(null);
        assertNotNull("value");

        // Boolean
        assertTrue(true);
        assertFalse(false);

        // Array
        assertArrayEquals(new int[]{1, 2, 3}, new int[]{1, 2, 3});

        // Floating point with delta
        assertEquals(0.1, 0.1001, 0.001);
    }

    // ── Summary ─────────────────────────────────────

    @Test
    void testSummary() {
        System.out.println("\n--- Summary ---");
        System.out.println("Java testing:");
        System.out.println("  - JUnit 5: @Test, @BeforeEach, @ParameterizedTest");
        System.out.println("  - Assertions: assertEquals, assertTrue, assertNull");
        System.out.println("  - Exception testing: assertThrows");
        System.out.println("  - Parameterized: @CsvSource, @ValueSource");
        System.out.println("  - Mocking: Mockito for object mocking");
        System.out.println("  - Lifecycle: @BeforeEach, @AfterEach");
        System.out.println("  - Coverage: JaCoCo, Cobertura");
    }
}
